File: VirtualLab_Python/SynDiag_Python/diagnostics/Tokalab/Diag_FluxLoops.py
# -*- coding: utf-8 -*-
"""
Created on Tue Jun  3 17:38:38 2025

@author: ricca
"""

import numpy as np
import os
from scipy.interpolate import griddata

class Diag_FluxLoops:
    def __init__(self):
        self.R = None  # Horizontal coordinate
        self.Z = None  # Vertical coordinate
        self.psi = None  # Measured magnetic flux
        self.unit = None  # Unit of measurement
        self.config = {}  # Configuration parameters
        self.ideal = {}  # Ideal (noise-free) measurements

    def measure(self, equi):
        
        points = np.column_stack((equi.geo.grid.Rg.flatten(), equi.geo.grid.Zg.flatten()))

        # Interpolazione con griddata (equiv. interp2 Matlab)
        psi = griddata(points, equi.psi.flatten(), (self.R, self.Z), method='linear')

        # Save ideal (noise-free) signal
        self.ideal["psi"] = psi
        
        # Add noise
        noise_abs = np.random.normal(0, self.config.get("noise_random_absolute_intensity", 0), size=np.shape(psi.ravel()))
        noise_prop = np.random.normal(0, np.abs(psi) * self.config.get("noise_random_proportional_intensity", 0))

        self.psi = psi + noise_abs + noise_prop
        self.unit = "Wb/rad"

    def upload(self, configuration=1):
        if configuration == 1:
            self.config["configuration"] = 1
            from scipy.io import loadmat
            module_path = os.path.abspath(__file__)
            module_path = os.path.dirname(module_path)
            module_path = os.path.join(module_path,"diagnostics_data","FluxLoopsData_config_1.mat")
            data = loadmat(module_path)
            
            self.R = np.squeeze(data["R"])
            self.Z = np.squeeze(data["Z"])

            self.config["noise_random_absolute_intensity"] = 0
            self.config["noise_random_proportional_intensity"] = 0

File: VirtualLab_MatLab/SynDiag_MATLAB/diagnostics/Tokalab/Diag_FluxLoops.m
%% tokalab diagnostics

classdef Diag_FluxLoops

    properties

        R % Horixontal coordinate
        Z % Vertical coordinate

        psi % Measured Flux

        unit % Unit Measure

        config % contains the various information such as noise, etc.

        ideal % contains the measurements without the noise

    end

    methods

        function obj = measure(obj,equi)

            R_equi = equi.geo.grid.Rg;
            Z_equi = equi.geo.grid.Zg;

            psi_equi = equi.psi;

            psi = interp2(R_equi,Z_equi,psi_equi,obj.R,obj.Z);

            obj.ideal.psi = psi;

            % noise absolute
            noise_abs = normrnd(0,obj.config.noise_random_absolute_intensity,size(obj.ideal.psi));

            % noise proportional
            noise_prop = normrnd(0,abs(obj.ideal.psi).*obj.config.noise_random_proportional_intensity);

            % real measurement
            obj.psi = obj.ideal.psi + noise_abs + noise_prop;

            obj.unit = "Wb/rad";

        end

        function obj = Upload(obj,configuration)

            % default configuration
            if nargin<2
                configuration = 1;
            end

            if configuration == 1

                obj.config.configuration = 1;

                load("diagnostics_data/FluxLoopsData_config_1.mat")

                obj.R = R;
                obj.Z = Z;

                obj.config.noise_random_absolute_intensity = 0;
                obj.config.noise_random_proportional_intensity = 0;

            end

        end


        %% Plotting Functions

        function plot_geo(obj)

            plot(obj.R,obj.Z,'.','MarkerSize',16)
            grid on
            grid minor
            xlabel("R")
            ylabel("Z")

        end

        function plot_meas(obj)

            plot(obj.psi,'.','MarkerSize',16)
            grid on
            grid minor
            xlabel("#")
            ylabel("\psi [Wb/rad]")

        end

        function plot_StandAlone(obj)

            hold off
            plot(obj.ideal.psi,'.b','MarkerSize',16)
            hold on
            plot(obj.psi,'or','LineWidth',1.2)
            grid on
            grid minor
            xlabel("#")
            ylabel("\psi [Wb/rad]")

        end

    end

end




